### ISSUE TITLE
Proposed new section - Design patterns

### ISSUE THREAD

**Issue by braydie** [braydie/HowToBeAProgrammer#33]
Currently we have 1.11 "How to Learn Design Skills", and I think getting an understanding of well-designed software is very important (especially as a beginner), but should we look to include a section later on (perhaps in Intermediate) on when to use (and not use) design patterns? Currently 1.11 makes no mention of them and even having an awareness of them can be beneficial.

**Comment by amrFHamza** [comment]
I agree with that

**Comment by oktapodia** [comment]
:+1:

**Comment by maciej-gurban** [comment]
Agree, however the topic seems so vast that it would deserve not only a chapter, but probably a whole book to at least lay out the types of methodologies and what each entails.

**Comment by braydie** [comment]
@maciej-gurban - completely agree with you. My guess would be that the chapter would merely introduce the concept of design patterns, summise the pros and cons of using them and then point to further reading on specifics

**Comment by oktapodia** [comment]
If you want to get a list of design pattern in PHP, you can go [here](https://github.com/domnikl/DesignPatternsPHP)

**Comment by RobertLRead** [comment]
I think we could definitely expand this section, with references, starting with the "Gang of Four" book and then reference design pattern books in particular languages.

As @maciej-gurban mentions, it is probably not best to try to teach design patterns here.  However, I would like to have a statement about how to recognize when a design pattern might help you (e.g., when you find yourself doing the same thing twice or three times)

**Comment by ImDineshSaini** [comment]
Software Architecture should be correct name to represent the section, we can add the small section inside that like: Design Pattern, Architecture Pattern

**Comment by braydie** [comment]
I've started something [here](https://github.com/braydie/HowToBeAProgrammer/blob/section-design-patterns/2-Intermediate/Judgment/09-Design%20Patterns.md)

It's in a separate branch and will remain so until we're happy with the content to go back into `master`

**Comment by andela-ijubril** [comment]
For design patterns in python https://github.com/faif/python-patterns

**Comment by ImDineshSaini** [comment]
Any suitable link for design pattern in ruby

**Comment by unknownbreaker** [comment]
How's this one going?
